### PR TITLE
Design Doc for PaddlePaddle TrainingJob k8s CRD

### PR DESCRIPTION
**Pull request by qizheng09**
hi, @typhoonzero @Yancey1989 ,  I'd like you to review this issue. In this issue. This is the design doc for paddlepaddle cluster job.

### CODE REVIEW

File: go/pkg/trainingjober/trainingjober.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
+// trainingjober is a package for managing a TrainingJob
```
Comment (Yancey1989) [inline go/pkg/trainingjober/trainingjober.go]: Sorry I don't think `trainingjober` is a suitable name, we need a *Controller* to manager the TrainingJob, so how about rename `TrainingJober` to `TrainingJobController` ?

File: go/pkg/trainingjober/trainingjober.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+const (
+	trainingJobEventDelete trainingJobEventType = "Delete"
+	trainingJobEventModify trainingJobEventType = "Modify"
+)
+
+type trainingJobEvent struct {
+	// pet is the TrainingJobEventType of TrainingJob
+	pet trainingJobEventType
+	// The job transfer the information fo job
```
Comment (Yancey1989) [inline go/pkg/trainingjober/trainingjober.go]: `The job transfer the information fo job `, Do you mean `job is the reference to TrainingJob` ?

File: go/pkg/trainingjober/trainingjober.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	job *v1alpha1.TrainingJob
+}
+
+// TrainingJober is to manager a specific TrainingJob
```
Comment (Yancey1989) [inline go/pkg/trainingjober/trainingjober.go]: I think we need a `TrainingjobController` to manager a TrainingJob.

**Comment by qizheng09** [comment]
@Yancey1989   We already have a controller which @m3ngyang committed I will write  a design doc to describe the definition of Controller、TrainingJober、TrainingJob and the state machine of a TrainingJob.

File: go/pkg/trainingjober/trainingjober.go (hunk 3 above)
Comment (qizheng09) [reply go/pkg/trainingjober/trainingjober.go]: We have a controller which @m3ngyang committed in another PR https://github.com/PaddlePaddle/cloud/pull/565. We use controller to manager all TrainingJob of a specific cluster and use TrainingJober to manager a specific TrainingJob.

File: go/pkg/design_doc.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
+# Design Doc for PaddlePaddle TrainingJob k8s CRD
```
Comment (typhoonzero) [inline go/pkg/design_doc.md]: Design docs must put under `docs/design`

Comment (qizheng09) [reply go/pkg/design_doc.md]: Thanks for your advice, I will put it under docs/design.

File: doc/design/traingjob_crd_design.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+- It will register TrainingJob CRD to cluster if there is no TrainingJob resource.
+- It will create a TrainingJober while PaddlePaddle job was submitted and delete the TrainingJober while job was 
+deleted.
+
+### Overall design
+
+The whole life cycle of TrainingJob is managed through the two layer control of Controller and TrainingJober. As 
+shown in the following figure:
+
+![image](https://github.com/qizheng09/figure/blob/master/paddlecloud/overall_design.png?raw=true)
```
Comment (typhoonzero) [inline doc/design/traingjob_crd_design.md]: Put the image in this repo. And, if you are using tools like "OmniGraffle" or something like that, you should put the source file in the repo too, so we can modify the graph later for updates.

Comment (typhoonzero) [reply doc/design/traingjob_crd_design.md]: This figure should have some explanation of how it runs and what components it consists of.

File: doc/design/traingjob_crd_design.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+The TrainingJob CRD defines a PaddlePaddle job resource for k8s, including cluster training parameters and a set of 
+PaddlePaddle roles such as master, pserver and trainer.
+
+- master is a k8s `ReplicaSet`
+- pserver is a k8s `ReplicaSet`
+- trainer is a k8s `job`
+
+Here is an example yaml for a PaddlePaddle cluster job.
+ 
+```yaml
```
Comment (typhoonzero) [inline doc/design/traingjob_crd_design.md]: Can you please try to merge this to current design: https://github.com/PaddlePaddle/cloud/blob/develop/doc/edl/README.md, so we only have one total design here.

File: doc/design/traingjob_crd_design.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+kind: TrainingJob
+metadata:
+  name: paddlejob
+  namespace: testspace
+spec:
+  #paddle image with training script
+  image: "paddlepaddle/paddlecloud-job"
+  # base port of pserver
+  port: 7164
+  # ports num default 1
```
Comment (typhoonzero) [inline doc/design/traingjob_crd_design.md]: These comments seem not necessary, we can explain the details of the yaml file in another markdown file.

File: doc/design/traingjob_crd_design.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        memory: "200Mi"
+      requests:
+        cpu: "200m"
+        memory: "200Mi"
+
+```                                                                                         
+
+The example will create a PaddlePaddle cluster job with a master, 2 pserver and 2 trainer.
+
+### TrainingJober
```
Comment (typhoonzero) [inline doc/design/traingjob_crd_design.md]: `TrainingJober` is a confusing name.

**Comment by m3ngyang** [comment]
too many commits for one pr, you'd better rebase and merge the commits.

File: doc/edl/README.md [hunk 9]
```diff
@@ -3,16 +3,30 @@
 ## Background
 
 A PaddlePaddle training job contains several trainer instances,
-several parameter server instances, and one master instance. We would
-like to automatically scale the number of training instances and the
+several parameter server instances, and one master instance.
```
Comment (typhoonzero) [inline doc/edl/README.md]: Background section should only tell us about "Why we need to do this", designs and thoughts need to put to bottom sections.

File: doc/edl/README.md [hunk 10]
```diff
@@ -3,16 +3,30 @@
...
+Currently, we have two controller Lifecycle Controller and Autoscale Controller to support the functions mentioned 
+above. We will merge them in the near future.
+
+**The goals of Lifecycle Controller including:**
+
+- Implement a controller and register a TrainingJob CRD
+- Create and delete TrainingJob
+- TrainingJob life cycle management
+
+**The goals of Autoscale Controller including:**
```
Comment (typhoonzero) [inline doc/edl/README.md]: We can call the controller "controller", and the `Autoscaler` is not a Kubernetes controller actually, It's just a goroutine running at the background to do autoscaling.

File: doc/edl/README.md [hunk 11]
```diff
@@ -58,45 +66,55 @@ describes the training job.
 A pseudo resource declaration (`training_job.yaml`) is as follows:
 
 ```yaml
-apiVersion: paddlepaddle.org/v1
+apiVersion: paddlepaddle.org/v1alpha1
```
Comment (typhoonzero) [inline doc/edl/README.md]: We can use `paddlepaddle.org/v1` so some code won't need big changes.

File: doc/edl/README.md [hunk 12]
```diff
@@ -107,60 +125,177 @@ optional: the master server will be created only when configured.
...
+machine of a TrainingJob:
 
-- [Custom Resource Definition (CRD)](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/),
-  since Kubernetes v1.7.
-- [Third Party Resource (TPR)](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-third-party-resource/),
-  since Kubernetes v1.2, fully deprecated in v1.8, will be removed in
-  v1.9.
+![](state_machine.jpg)
 
-We will support TPR first, because some of our clusters is using
```
Comment (typhoonzero) [inline doc/edl/README.md]: We may still need this line somewhere.

**Comment by qizheng09** [comment]
@typhoonzero , I fix the comments  you mentioned above. How do you think about this commit?

File: doc/edl/README.md [hunk 13]
```diff
@@ -61,103 +58,170 @@ A pseudo resource declaration (`training_job.yaml`) is as follows:
...
+// The job transfer the information fo job
+job *v1.TrainingJob
+}
 
-The training job custom resource can be created with: `kubectl create
--f training_job.yaml`.
+// TrainingJobUpdater is to manager a specific TrainingJob
+type TrainingJobUpdater struct {
+// job is the job the TrainingJobUpdater manager.
+job *v1.TrainingJob
```
Comment (m3ngyang) [inline doc/edl/README.md]: code format problem

**Pull request merged**